cpp/src/io/comp/compression.cpp

```diff
@@ -325,73 +327,100 @@ void host_compress(compression_type compression,
     h_results[idx]                  = {bytes_written, codec_status::SUCCESS};
   }
+
   cudf::detail::cuda_memcpy<codec_exec_result>(results, h_results, stream);
 }
 
-[[nodiscard]] bool use_host_compression(
-  compression_type compression,
-  [[maybe_unused]] device_span<device_span<uint8_t const> const> inputs,
-  [[maybe_unused]] device_span<device_span<uint8_t> const> outputs)
+enum class host_engine_state : uint8_t { ON, OFF, AUTO, HYBRID };
+
+[[nodiscard]] host_engine_state get_host_engine_state(compression_type compression)
 {
   auto const has_host_support   = is_host_compression_supported(compression);
   auto const has_device_support = is_device_compression_supported(compression);
   CUDF_EXPECTS(has_host_support or has_device_support,
                "Unsupported compression type: " + compression_type_name(compression));
-  if (not has_host_support) { return false; }
-  if (not has_device_support) { return true; }
-  // If both host and device compression are supported, dispatch based on the environment variable
+  if (not has_host_support) { return host_engine_state::OFF; }
+  if (not has_device_support) { return host_engine_state::ON; }
 
+  // If both host and device compression are supported, dispatch based on the environment variable
   auto const env_var = getenv_or("LIBCUDF_HOST_COMPRESSION", std::string{"OFF"});
+
   if (env_var == "AUTO") {
+    return host_engine_state::AUTO;
+  } else if (env_var == "HYBRID") {
+    return host_engine_state::HYBRID;
+  } else if (env_var == "OFF") {
+    return host_engine_state::OFF;
+  } else if (env_var == "ON") {
+    return host_engine_state::ON;
+  }
+  CUDF_FAIL("Invalid LIBCUDF_HOST_COMPRESSION value: " + env_var);
+}
+
+[[nodiscard]] size_t find_split_index(device_span<device_span<uint8_t const> const> inputs,
+                                      host_engine_state host_state,
+                                      rmm::cuda_stream_view stream)
+{
+  CUDF_FUNC_RANGE();
+  if (host_state == host_engine_state::OFF or inputs.empty()) { return 0; }
+  if (host_state == host_engine_state::ON) { return inputs.size(); }
+
+  if (host_state == host_engine_state::AUTO) {
     auto const threshold =
       getenv_or("LIBCUDF_HOST_COMPRESSION_THRESHOLD", default_host_compression_auto_threshold);
-    return inputs.size() < threshold;
+    return inputs.size() < threshold ? inputs.size() : 0;
   }
 
-  return env_var == "ON";
+  if (host_state == host_engine_state::HYBRID) {
+    auto const h_inputs    = cudf::detail::make_host_vector(inputs, stream);
+    size_t total_host_size = 0;
+    size_t total_host_to_single_device_thread_ratio =
+      getenv_or("LIBCUDF_HOST_COMPRESSION_RATIO", default_host_compression_ratio);
+    for (size_t i = 0; i < h_inputs.size(); ++i) {
+      if (total_host_size >= total_host_to_single_device_thread_ratio * h_inputs[i].size()) {
+        return i;
+      }
+      total_host_size += h_inputs[i].size();
+    }
```

Comment:
Loop could use modernization

```suggestion
    for (auto const& h_input : h_inputs) {
      if (total_host_size >= total_host_to_single_device_thread_ratio * h_input.size()) {
        return i;
      }
      total_host_size += h_input.size();
    }
```